### PR TITLE
NEW: Start adding Collection

### PR DESCRIPTION
**Pull request by Oddant1**
Adds new Collection type to support collections of outputs

### CODE REVIEW

File: qiime2/core/testing/method.py
Comment (ebolyen) [inline qiime2/core/testing/method.py:143]: Forgot about how IntSequence works. I think the type here for a collection of them would actually be:
`typing.List[list]` (you can access the inner fields by using `.__args__` -> `(list,)`)


Alternatively, we may want to disallow list and dict in transformers, although it will be a huge pain to rewrite these test actions, so if it can be done without that, I'm all for it.

**Comment by Oddant1** [comment]
Blocking qiime2/q2cli#288

File: qiime2/core/cache.py
Comment (ebolyen) [inline qiime2/core/cache.py:773]: This could take a while, do we need to do something similar to `_rename_to_data` so that we aren't locking a cache for the duration it takes a collection to be (re-)written to a cache?

File: qiime2/core/cache.py
Comment (ebolyen) [inline qiime2/core/cache.py:821]: I was thinking we may want to set `default_flow_style=False`, but after testing it with a collection, it seems to work fine. :+1:

File: qiime2/core/cache.py
Comment (ebolyen) [inline qiime2/core/cache.py:893]: This must be for the CLI's benefit? I would generally expect an error, since the return type is dramatically different, but I see why it would be significantly simpler to just get _something_ out of the cache if it's real.

File: qiime2/core/type/collection.py
Comment (ebolyen) [inline qiime2/core/type/collection.py:130]:
```suggestion
```

`__iter__` is very special to the type system, so I would avoid implementing it here

File: qiime2/core/type/signature.py
Comment (ebolyen) [inline qiime2/core/type/signature.py:321]: (Note to self to pick up from here tomorrow.)

**Review by ebolyen — CHANGES_REQUESTED** [review]
Main thing I see so far is just the time that the cache may be locked when storing a collection (only really comes up in a cross-cache situation).

File: qiime2/core/type/signature.py
Comment (ebolyen) [inline qiime2/core/type/signature.py:321]: It would be nice if these layers did not have to consider provenance, that feels like something that should stay (be trapped) in `action.py` since provenance is quite mutable whereas things in `core.type` are quite functional.

In the case of parameters, you can pretty much do that without any change here, but the inputs and outputs seem trickier. So probably not worth addressing right now.

I really wish something like this was straight-forward:

```python
map(invocation.inputs, lambda name, spec, value: value._view(...)
```
where this "layer" basically just resolves the difference between lists/dicts/singular elements and transforms between them as needed. Or perhaps there's some dictionary of generic "Argument" objects which you can apply transformations to and it internally magics away the distinction.

**Review by ebolyen — CHANGES_REQUESTED** [review]
It's mostly just some really bizarre edge-cases at this point. Nothing strictly blocks this PR, but we will need to circle back to fix at some point before release.

File: qiime2/core/cache.py
Comment (Oddant1) [reply qiime2/core/cache.py:772]: Got so used to casting it as a list that I started doing it everywhere without thinking

File: qiime2/core/cache.py
Comment (Oddant1) [reply qiime2/core/cache.py:893]: Yes, this is for the CLI's benefit. I didn't really want it to work this way, but it makes the CLI so much simpler.

**Comment by lizgehret** [comment]
This all _generally_ makes sense at a high level. I'll do another pass after you've addressed all of Evan's review comments!

File: qiime2/core/cache.py
Comment (Oddant1) [reply qiime2/core/cache.py:773]: pool.save is already pretty efficient, it creates a symlink in the pool then calls Cache._rename_to_data

File: qiime2/core/cache.py
Comment (Oddant1) [reply qiime2/core/cache.py:893]: Follow up on this, should save dispatch to save_collection or error? It would be nice to keep this behavior consistent. I can also always make it so load doesn't dispatch to load_collection, but IIRC that was looking like it was going to add a bit on annoyance on the CLI, but should be completely doable.

Comment (Oddant1) [reply qiime2/core/cache.py:893]: I'm honestly inclined to just kinda have save/load defer to save/load collection. It is nice to make people be explicit but idk it just feels like quality of life in this instance

**Pull request merged**
